File: design-model/builder.js
/**
 * Design Builder — the loop that makes Cognitive actually build things.
 *
 * generate → screenshot → evaluate → iterate → deliver
 *
 * Any agent calls design.build(goal) → gets production-quality code
 * that passed design evaluation. Emotion state drives creative direction.
 *
 *   import { build } from './design-model/builder.js';
 *
 *   const result = await build({
 *     goal: 'A macOS settings panel for a photo editing app',
 *     style: 'dark, minimal, native',
 *     constraints: ['keyboard shortcuts', 'max 5 settings visible'],
 *     emotionState: oca.layers.emotion?.getState(),
 *   });
 *
 *   // result.code = '<html>...'
 *   // result.scores.overall_aesthetic = 0.82
 *   // result.iterations = 4
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, unlinkSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { execSync } from 'child_process';
import { tmpdir } from 'os';

import { evaluateDesign } from './evaluate.js';
import { suggestChanges, generateCssPatch } from './suggest.js';
import { capture, cleanup } from './screenshot-capture.js';
import { encodeFromCode } from './encoder.js';
import { SCORE_NAMES, DESIGN_DIMENSIONS } from './knowledge.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Try to import emotion bridge (optional — works without OCA)
let computeDesignPolicy, policyToPromptContext;
try {
  const bridge = await import('../design/emotion-bridge.js');
  computeDesignPolicy = bridge.computeDesignPolicy;
  policyToPromptContext = bridge.policyToPromptContext;
} catch {}

// Load the design skill for prompt assembly
function loadSkill() {
  const skillPath = process.env.HOME + '/.claude/skills/frontend-design/SKILL.md';
  if (existsSync(skillPath)) {
    return readFileSync(skillPath, 'utf-8');
  }
  return '';
}

// Default LLM call via Claude CLI
function defaultLlmCall(prompt) {
  const tmp = join(tmpdir(), `builder-${Date.now()}.txt`);
  writeFileSync(tmp, prompt);
  try {
    return execSync(`cat "${tmp}" | claude -p --model sonnet`, {
      encoding: 'utf-8', timeout: 600000, maxBuffer: 4 * 1024 * 1024, shell: '/bin/zsh',
    }).trim();
  } finally {
    try { unlinkSync(tmp); } catch {}
  }
}

// ═══════════════════════════════════════════════════
// PROMPT ASSEMBLY
// ═══════════════════════════════════════════════════

function assembleGenerationPrompt(goal, options = {}) {
  const { style, constraints, emotionState, skillContent } = options;

  let prompt = `You are building a real, production-quality interface. Not a mockup. Not a prototype. Something that ships.

DESIGN SKILL:
${(skillContent || '').slice(0, 2000)}

`;

  // Emotion-to-design policy
  if (emotionState && computeDesignPolicy) {
    try {
      const policy = computeDesignPolicy(emotionState);
      const context = policyToPromptContext(policy);
      prompt += `EMOTIONAL DESIGN CONTEXT:\n${context}\n\n`;
    } catch {}
  }

  prompt += `GOAL: ${goal}\n`;
  if (style) prompt += `STYLE DIRECTION: ${style}\n`;
  if (constraints?.length) prompt += `CONSTRAINTS:\n${constraints.map(c => `- ${c}`).join('\n')}\n`;

  prompt += `
REQUIREMENTS:
- Complete HTML page with inline CSS, viewport 1440x900
- System font stack (-apple-system, BlinkMacSystemFont, "SF Pro Text", system-ui)
- CSS custom properties for the entire color system
- Every design choice must have a reason. If you use a color, know why.
- The design must be COHESIVE — parts working together as one vision, not individually good but clashing
- Think about what makes this DISTINCTIVE. What's the one thing someone will remember?
- Innovation matters: solve the problem in a way that surprises. What would NotchNook or Klack do?

Return ONLY the complete HTML. No markdown, no explanation, no code fences.`;

  return prompt;
}

function assembleIterationPrompt(code, evaluation, suggestions) {
  const weakDims = evaluation.weakest.slice(0, 3);
  const weakFeedback = weakDims.map(w => {
    const dim = DESIGN_DIMENSIONS.find(d => d.name === w.name);
    return `- ${w.name}: ${(w.score || 0).toFixed(2)} — ${dim?.highSignals?.slice(0, 2).join(', ') || 'improve quality'}`;
  }).join('\n');

  const cssPatch = suggestions?.changes ? generateCssPatch(suggestions.changes.slice(0, 5)) : '';

  return `Improve this design. The following dimensions are weak:

${weakFeedback}

SPECIFIC CHANGES TO MAKE:
${suggestions?.summary || 'Improve the weakest dimensions listed above.'}

SUGGESTED CSS (incorporate these ideas, don't blindly paste):
${cssPatch.slice(0, 1000)}

CURRENT CODE:
${code.slice(0, 8000)}

CRITICAL: The improved version must be COHESIVE. Don't just fix one dimension — make sure your changes work with the WHOLE design. A great heading that fights the color palette makes the design WORSE.

Return ONLY the complete improved HTML. No markdown, no explanation.`;
}

// ═══════════════════════════════════════════════════
// BUILD LOOP
// ═══════════════════════════════════════════════════

/**
 * Build a design artifact with quality-guided iteration.
 *
 * @param {object} options
 * @param {string} options.goal - What to build
 * @param {string} [options.style] - Aesthetic direction
 * @param {string[]} [options.constraints] - Design constraints
 * @param {object} [options.emotionState] - OCA emotion state (PADCN + channels + drives)
 * @param {Function} [options.llmCall] - LLM call function (default: Claude CLI)
 * @param {number} [options.maxIterations=8] - Maximum iteration attempts
 * @param {number} [options.qualityThreshold=0.72] - Minimum overall_aesthetic to accept
 * @param {number} [options.innovationThreshold=0.4] - Minimum innovation_score to accept
 * @param {string} [options.renderer='html'] - 'html' (Puppeteer) or 'swift' (future)
 * @param {string} [options.outputDir] - Where to save artifacts
 * @returns {object} { code, screenshot, scores, iterations, history, suggestions, success }
 */
export async function build(options) {
  const {
    goal,
    style,
    constraints,
    emotionState,
    llmCall = defaultLlmCall,
    maxIterations = 8,
    qualityThreshold = 0.72,
    innovationThreshold = 0.4,
    renderer = 'html',
    outputDir,
  } = options;

  const buildDir = outputDir || join(__dirname, 'builds', `build-${Date.now()}`);
  mkdirSync(buildDir, { recursive: true });

  const skillContent = loadSkill();
  const history = [];
  let bestCode = null;
  let bestScore = 0;
  let bestScores = null;
  let bestScreenshot = null;

  console.log(`[builder] goal: ${goal}`);
  console.log(`[builder] thresholds: quality=${qualityThreshold} innovation=${innovationThreshold}`);

  for (let iteration = 0; iteration < maxIterations; iteration++) {
    const isFirst = iteration === 0;

    // ── GENERATE ──
    let code;
    try {
      const lastEntry = history.length > 0 ? history[history.length - 1] : null;
      if (!lastEntry || !lastEntry.code) {
        // First generation or retry after failure
        const prompt = assembleGenerationPrompt(goal, { style, constraints, emotionState, skillContent });
        code = llmCall(prompt);
      } else {
        // Iterate on previous version
        const prompt = assembleIterationPrompt(lastEntry.code, lastEntry.evaluation, lastEntry.suggestions);
        code = llmCall(prompt);
      }
      // Clean markdown fences
      const match = code?.match(/```html\n?([\s\S]*?)\n?```/);
      if (match) code = match[1];
      if (!code || !code.includes('<') || code.length < 100) {
        console.log(`  [${iteration + 1}] invalid HTML, skipping`);
        continue;
      }
    } catch (e) {
      console.log(`  [${iteration + 1}] generation failed: ${e.message.slice(0, 50)}`);
      continue;
    }

    // ── RENDER ──
    const htmlPath = join(buildDir, `iteration-${iteration}.html`);
    const pngPath = join(buildDir, `iteration-${iteration}.png`);
    writeFileSync(htmlPath, code);

    try {
      await capture(`file://${htmlPath}`, pngPath);
    } catch (e) {
      console.log(`  [${iteration + 1}] render failed: ${e.message.slice(0, 50)}`);
    }

    // ── EVALUATE ──
    const evalInput = { code };
    if (existsSync(pngPath)) evalInput.screenshot = pngPath;
    const evaluation = await evaluateDesign(evalInput);
    const overall = evaluation.overall || 0;
    const innovation = evaluation.scores?.innovation_score || 0;

    // ── SUGGEST ──
    let suggestions = null;
    try {
      suggestions = await suggestChanges(code, existsSync(pngPath) ? pngPath : undefined);
    } catch {}

    // Track
    history.push({
      iteration,
      code,
      htmlPath,
      screenshotPath: existsSync(pngPath) ? pngPath : null,
      evaluation,
      suggestions,
      scores: evaluation.scores,
      overall,
      innovation,
    });

    // Track best
    if (overall > bestScore) {
      bestCode = code;
      bestScore = overall;
      bestScores = evaluation.scores;
      bestScreenshot = existsSync(pngPath) ? pngPath : null;
    }

    const marker = overall >= qualityThreshold && innovation >= innovationThreshold ? '✅' :
                   overall >= qualityThreshold ? '🟡' : '⏳';
    console.log(`  [${iteration + 1}/${maxIterations}] overall=${overall.toFixed(3)} innovation=${innovation.toFixed(3)} ${marker}`);

    // ── DECIDE ──
    if (overall >= qualityThreshold && innovation >= innovationThreshold) {
      console.log(`[builder] accepted at iteration ${iteration + 1}`);
      break;
    }

    if (iteration === maxIterations - 1) {
      console.log(`[builder] max iterations reached, using best (${bestScore.toFixed(3)})`);
    }
  }

  await cleanup(); // Close Puppeteer

  // ── STORE AS TRAINING DATA ──
  try {
    const manifestPath = join(__dirname, 'data', 'manifest.json');
    if (existsSync(manifestPath) && bestScreenshot) {
      const manifest = JSON.parse(readFileSync(manifestPath, 'utf-8'));
      let codeFeatures = null;
      try { codeFeatures = Array.from(encodeFromCode(bestCode, { platform: 'web' })); } catch {}

      manifest.samples.push({
        image: bestScreenshot,
        screenshot_path: bestScreenshot,
        scores: bestScores,
        source: 'builder',
        code_features: codeFeatures,
        metadata: {
          source_name: `build-${goal.slice(0, 30).replace(/\s+/g, '-')}`,
          category: 'builder-output',
          quality_target: bestScore > 0.8 ? 'high' : 'medium',
          goal: goal.slice(0, 200),
          iterations: history.length,
          code_path: history[history.length - 1]?.htmlPath,
          generated_at: new Date().toISOString(),
        },
      });
      writeFileSync(manifestPath, JSON.stringify(manifest, null, 2));
    }
  } catch {}

  return {
    code: bestCode,
    screenshot: bestScreenshot,
    scores: bestScores,
    overall: bestScore,
    iterations: history.length,
    history: history.map(h => ({
      iteration: h.iteration,
      overall: h.overall,
      innovation: h.innovation,
      screenshotPath: h.screenshotPath,
    })),
    suggestions: history[history.length - 1]?.suggestions,
    success: bestScore >= qualityThreshold,
    buildDir,
  };
}

export default { build };

File: design-model/builder.py
#!/usr/bin/env python3
"""
Design Builder — generate → render → evaluate → iterate → deliver.

Uses the Anthropic API directly. No CLI hangs.
Language-aware: produces HTML, SwiftUI, or React.  Accretes into a
persistent project dir when --project is passed (so OCA builds a single
app over time instead of scattering throwaway builds).

Usage:
  python builder.py "A focus timer" --language swiftui
  python builder.py "A settings panel" --style "dark, native" --iterations 6
  python builder.py "Next iteration" --project presence   # accretes into active-project/presence/iterations/
  python builder.py --help
"""

import argparse
import json
import os
import re
import subprocess
import sys
import time
from pathlib import Path

import anthropic
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent.parent / ".env", override=True)

client = anthropic.Anthropic()

# Reuse self_train.py's language-aware helpers so builder + self-train stay
# in sync on fence extraction, prompt shape, and rendering.
sys.path.insert(0, str(Path(__file__).parent))
from self_train import LANGUAGE_PROMPTS, extract_code  # noqa: E402
from renderers import render as render_code  # noqa: E402

SCORE_NAMES = [
    "typography_quality", "color_harmony", "spatial_composition",
    "motion_elegance", "emotional_resonance", "craft_visibility",
    "minimalism_coherence", "native_integration",
    "visceral_score", "behavioral_score", "reflective_score",
    "overall_aesthetic",
    "innovation_score", "system_creativity", "design_distinctiveness",
    "problem_level",
]

LANGUAGE_EXT = {"html": "html", "swiftui": "swift", "react": "jsx"}

BUILDS_DIR = Path(__file__).parent / "builds"
ACTIVE_PROJECT_DIR = Path(__file__).parent / "active-project"
MANIFEST_PATH = Path(__file__).parent / "data" / "manifest.json"
SKILL_PATH = Path.home() / ".claude" / "skills" / "frontend-design" / "SKILL.md"


def sonnet(prompt: str) -> str:
    msg = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=16000,
        messages=[{"role": "user", "content": prompt}],
    )
    return msg.content[0].text


def opus(prompt: str) -> str:
    msg = client.messages.create(
        model="claude-opus-4-20250514",
        max_tokens=4000,
        messages=[{"role": "user", "content": prompt}],
    )
    return msg.content[0].text


def render_to_png(code: str, png_path: str, language: str) -> bool:
    """Render the generated code to a screenshot via renderers.py dispatch."""
    try:
        success, out_path, err = render_code(language, code, png_path)
        if not success:
            print(f"  [render] {language} failed: {(err or '').strip()[:120]}")
        return success
    except Exception as e:
        print(f"  [render] exception: {e}")
        return False


def grade(code: str, language: str) -> dict | None:
    """Opus grades the design. Prompt is language-aware so the judge knows
    it's looking at SwiftUI / React / HTML and calibrates native_integration."""
    try:
        lang_label = {
            "html": "HTML/CSS web component",
            "swiftui": "native SwiftUI macOS view",
            "react": "React component",
        }.get(language, "UI design")

        raw = opus(
            f"Grade this {lang_label} 0-1 per dimension. Be HARSH and COHERENT — "
            f"great typography that clashes with the palette means BOTH drop.\n\n"
            f"For native_integration: {'reward platform-native idioms (SF Pro, NSVisualEffectView, system materials)' if language == 'swiftui' else 'reward fit-for-web conventions'}.\n\n"
            f"Dimensions: {', '.join(SCORE_NAMES)}\n\n"
            f"2-sentence critique. What's weak and HOW to fix it.\n"
            f'Return ONLY JSON: {{"scores":{{...all 16...}},"critique":"...","fixes":["specific fix 1","specific fix 2","specific fix 3"]}}\n\n'
            f"{language.upper()}:\n{code[:8000]}"
        )
        m = re.search(r"\{[\s\S]*\}", raw)
        if not m:
            return None
        result = json.loads(m.group())
        for name in SCORE_NAMES:
            result["scores"][name] = max(0.0, min(1.0, float(result["scores"].get(name, 0.5))))
        return result
    except Exception as e:
        print(f"  [grade failed] {e}")
        return None


def _next_iter_num(project_dir: Path) -> int:
    """Find the next iteration number in active-project/<name>/iterations/."""
    iters_dir = project_dir / "iterations"
    if not iters_dir.exists():
        return 1
    existing = [p for p in iters_dir.iterdir() if p.is_dir() and p.name.startswith("iter-")]
    if not existing:
        return 1
    nums = []
    for p in existing:
        try:
            nums.append(int(p.name.split("-", 1)[1]))
        except Exception:
            continue
    return (max(nums) + 1) if nums else 1


def build(
    goal: str,
    style: str = "",
    constraints: list[str] = None,
    max_iterations: int = 6,
    quality_threshold: float = 0.72,
    innovation_threshold: float = 0.40,
    language: str = "swiftui",
    project: str | None = None,
) -> dict:
    """
    Build a design artifact with quality-guided iteration.

    If `project` is given, accretes into active-project/<project>/iterations/iter-<N>/
    (the singular-project pattern — OCA iterates on one app over time).
    Otherwise writes a throwaway into builds/build-<timestamp>/.

    Returns: {code, screenshot, scores, iterations, history, success, build_dir, language}
    """
    if language not in LANGUAGE_PROMPTS:
        raise ValueError(f"Unknown language: {language}. Expected one of {list(LANGUAGE_PROMPTS)}")

    # ── Build-dir selection ──
    if project:
        project_dir = ACTIVE_PROJECT_DIR / project
        iter_num = _next_iter_num(project_dir)
        build_dir = project_dir / "iterations" / f"iter-{iter_num:04d}"
        print(f"[builder] project: {project} — iteration {iter_num}")
    else:
        build_dir = BUILDS_DIR / f"build-{int(time.time())}"
    build_dir.mkdir(parents=True, exist_ok=True)

    ext = LANGUAGE_EXT[language]
    lang_cfg = LANGUAGE_PROMPTS[language]

    # Load skill
    skill = SKILL_PATH.read_text()[:1500] if SKILL_PATH.exists() else ""

    print(f"[builder] goal: {goal}")
    print(f"[builder] language: {language} | thresholds: quality={quality_threshold} innovation={innovation_threshold}")

    history = []
    best_code = None
    best_score = 0
    best_scores = None
    best_png = None

    for i in range(max_iterations):
        t0 = time.time()

        # ── GENERATE ──
        if i == 0 or not history:
            parts = []
            if skill:
                parts.append(f"DESIGN SKILL:\n{skill}\n")
            parts.append(f"{lang_cfg['framing']}\n\nGoal: {goal}")
            if style:
                parts.append(f"Style: {style}")
            if constraints:
                parts.append("Constraints:\n" + "\n".join(f"- {c}" for c in constraints))
            parts.append(f"Rules:\n{lang_cfg['rules']}")
            parts.append(
                "Aim for Level 5+ thinking (problem redefinition, not just execution). "
                "Cohesive design — every part must serve the whole."
            )
            parts.append(lang_cfg["format"])
            prompt = "\n\n".join(parts)
        else:
            last = history[-1]
            fixes = last.get("fixes", [])
            prompt = (
                f"Improve this {language} design. Current score: {last['overall']:.2f}\n\n"
                f"Critique: {last.get('critique', '')}\n\n"
                f"Specific fixes needed:\n" +
                "\n".join(f"- {f}" for f in fixes[:5]) +
                f"\n\nCurrent code:\n{last['code'][:6000]}\n\n"
                f"Return the COMPLETE improved code. {lang_cfg['format']}\n"
                f"Make sure fixes work with the WHOLE design."
            )

        try:
            raw = sonnet(prompt)
            code = extract_code(raw, language)
            if len(code) < 100:
                print(f"  [{i+1}] generated code too short, skipping")
                continue
        except Exception as e:
            print(f"  [{i+1}] generate failed: {e}")
            continue

        gen_time = time.time() - t0

        # ── RENDER ──
        code_path = build_dir / f"iter-{i}.{ext}"
        png_path = build_dir / f"iter-{i}.png"
        code_path.write_text(code)
        has_png = render_to_png(code, str(png_path), language)

        # ── GRADE ──
        t1 = time.time()
        result = grade(code, language)
        grade_time = time.time() - t1

        if not result:
            print(f"  [{i+1}] grading failed, skipping")
            continue

        overall = result["scores"]["overall_aesthetic"]
        innovation = result["scores"]["innovation_score"]
        critique = result.get("critique", "")
        fixes = result.get("fixes", [])

        history.append({
            "iteration": i,
            "code": code,
            "code_path": str(code_path),
            "png_path": str(png_path) if has_png else None,
            "overall": overall,
            "innovation": innovation,
            "scores": result["scores"],
            "critique": critique,
            "fixes": fixes,
        })

        if overall > best_score:
            best_code = code
            best_score = overall
            best_scores = result["scores"]
            best_png = str(png_path) if has_png else None

        marker = "✅" if overall >= quality_threshold and innovation >= innovation_threshold else "⏳"
        print(f"  [{i+1}/{max_iterations}] overall={overall:.2f} innovation={innovation:.2f} "
              f"({gen_time:.0f}s gen, {grade_time:.0f}s grade) {marker}")
        if critique:
            print(f"    {critique[:80]}")

        # ── DECIDE ──
        if overall >= quality_threshold and innovation >= innovation_threshold:
            print(f"[builder] ✅ accepted at iteration {i+1}")
            break

    # ── STORE ──
    if best_code and best_png:
        try:
            manifest = json.loads(MANIFEST_PATH.read_text())
            manifest["samples"].append({
                "image": best_png,
                "screenshot_path": best_png,
                "scores": best_scores,
                "source": "builder",
                "confidence": 0.9,
                "metadata": {
                    "source_name": f"build-{goal[:30].replace(' ', '-')}",
                    "category": "builder-output",
                    "language": language,
                    "project": project or None,
                    "quality_target": "high" if best_score > 0.75 else "medium",
                    "goal": goal[:200],
                    "iterations": len(history),
                    "generated_at": time.strftime("%Y-%m-%dT%H:%M:%S"),
                },
            })
            MANIFEST_PATH.write_text(json.dumps(manifest, indent=2))
        except Exception:
            pass

    return {
        "code": best_code,
        "screenshot": best_png,
        "scores": best_scores,
        "overall": best_score,
        "iterations": len(history),
        "language": language,
        "project": project,
        "history": [{"iter": h["iteration"], "overall": h["overall"], "innovation": h["innovation"]} for h in history],
        "success": best_score >= quality_threshold,
        "build_dir": str(build_dir),
    }


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Design Builder — generate, evaluate, iterate")
    parser.add_argument("goal", help="What to build")
    parser.add_argument("--style", default="", help="Aesthetic direction")
    parser.add_argument("--constraints", nargs="*", default=[], help="Design constraints")
    parser.add_argument("--iterations", type=int, default=6, help="Max iterations")
    parser.add_argument("--quality", type=float, default=0.72, help="Quality threshold")
    parser.add_argument("--innovation", type=float, default=0.40, help="Innovation threshold")
    parser.add_argument(
        "--language",
        choices=["html", "swiftui", "react"],
        default="swiftui",
        help="Target language for generation (default: swiftui)",
    )
    parser.add_argument(
        "--project",
        default=None,
        help="Accrete into active-project/<project>/iterations/ instead of builds/<timestamp>/",
    )
    args = parser.parse_args()

    result = build(
        goal=args.goal,
        style=args.style,
        constraints=args.constraints,
        max_iterations=args.iterations,
        quality_threshold=args.quality,
        innovation_threshold=args.innovation,
        language=args.language,
        project=args.project,
    )

    ext = LANGUAGE_EXT[result["language"]]
    print(f"\n{'═' * 50}")
    print(f"{'SUCCESS' if result['success'] else 'BEST EFFORT'}")
    print(f"Language: {result['language']}")
    if result['project']:
        print(f"Project: {result['project']}")
    print(f"Overall: {result['overall']:.3f}")
    print(f"Iterations: {result['iterations']}")
    trajectory = ' → '.join(str(round(h["overall"], 2)) for h in result['history'])
    print(f"Trajectory: {trajectory}")
    print(f"Build: {result['build_dir']}")
    last_idx = max(0, result['iterations'] - 1)
    print(f"Open: {result['build_dir']}/iter-{last_idx}.{ext}")
    print("═" * 50)
